Ilariae/token bridge sdk (#96)
wormhole-foundation/wormhole-docs: build/contract-integrations/token-bridge.md

Reviewer: Is adding a card for Token Bridge on /build/contract-integrations/ index page part of this PR or will it be done separately?

Behind the scenes, when `ITokenBridge.attestToken()` is called with a given token address, the Token Bridge contract:

- Calls the token contract’s `decimals()`, `symbol()`, and `name()` functions
- Uses these values to create a metadata payload (`payload_id` set to `2`) that describes the token
- Emits a VAA containing this metadata, which the Guardians sign and publish

Under the hood, calling `tokenBridge.createAttestation()` uses Wormhole’s core method:

- `ITokenBridge.attestToken()` from [`src/interfaces/ITokenBridge.sol`](https://github.com/wormhole-foundation/wormhole-solidity-sdk/blob/main/src/interfaces/ITokenBridge.sol){target=\_blank}
- The logic for creating the attestation VAA can be found in `bridge/Bridge.sol`, specifically in the [`attestToken`](https://github.com/wormhole-foundation/wormhole/blob/main/ethereum/contracts/bridge/Bridge.sol#L38){target=\_blank} function

```ts
// In your code, you might have a tokenBridge instance set up from tokenBridge.ts
// This references the Wormhole SDK that internally calls `ITokenBridge.attestToken`.

> const tokenAddress = "<YOUR_TOKEN_ADDRESS>";

Reviewer: ```suggestion
const tokenAddress = INSERT_YOUR_TOKEN_ADDRESS;
```

for await (const tx of tokenBridge.createAttestation(tokenAddress)) {
  // Sign and send the transaction (e.g., via your wallet or ethers.js)
  await sendTransaction(tx);
}

// Once Wormhole Guardians generate the attestation VAA, you'll submit it on the target chain:
const attestationVAA = ... // obtain from Wormhole Guardian network
for await (const tx of tokenBridge.submitAttestation(attestationVAA)) {
  await sendTransaction(tx);
}
```

- The `createAttestation` method is defined in the Wormhole SDK's [`TokenBridge` interface](https://github.com/wormhole-foundation/wormhole-sdk-ts/blob/main/core/definitions/src/protocols/tokenBridge/tokenBridge.ts#L188){target=\_blank}
- On-chain, the `attestToken` method can be found in `bridge/Bridge.sol` and is part of the [`ITokenBridge` interface](https://github.com/wormhole-foundation/wormhole/blob/main/ethereum/contracts/bridge/Bridge.sol#L38){target=\_blank}

!!!important

>     - Ensure the token contract on the source chain implements standard ERC-20 metadata functions (`decimals()`, `symbol()`, `name()`).

Reviewer: ```suggestion
    - Ensure the token contract on the source chain implements standard ERC-20 metadata functions (`decimals()`, `symbol()`, `name()`)
```

>     - Call `attestToken()` (via the Wormhole SDK or directly on the contract) for the token address.

Reviewer: ```suggestion
    - Call `attestToken()` (via the Wormhole SDK or directly on the contract) for the token address
```

>     - You don't have to put the token details anywhere in your code. If the token contract is a standard ERC-20, the Token Bridge will read its metadata.

Reviewer: ```suggestion
    - You don't have to put the token details anywhere in your code. If the token contract is a standard ERC-20, the Token Bridge will read its metadata
```

>     - The attestation may fail or produce incomplete metadata if the token does not implement these standard functions. In that case, you must ensure the token is ERC-20 compliant.

Reviewer: ```suggestion
    - The attestation may fail or produce incomplete metadata if the token does not implement these standard functions. In that case, you must ensure the token is ERC-20 compliant
```

- **Destination chain redemption** - `ITokenBridge.completeTransfer()` is also defined in [`ITokenBridge.sol`](https://github.com/wormhole-foundation/wormhole-solidity-sdk/blob/main/src/interfaces/ITokenBridge.sol#L120){target=\_blank} and implemented in [`Bridge.sol`](https://github.com/wormhole-foundation/wormhole/blob/main/ethereum/contracts/bridge/Bridge.sol#L468){target=\_blank}. It verifies the VAA and mints or releases the tokens
- **SDK integration** - The Wormhole SDK provides convenient methods like [`tokenBridge.transfer()`](https://github.com/wormhole-foundation/wormhole-sdk-ts/blob/main/core/definitions/src/protocols/tokenBridge/tokenBridge.ts#L215){target=\_blank} and [`tokenBridge.redeem()`](https://github.com/wormhole-foundation/wormhole-sdk-ts/blob/main/core/definitions/src/protocols/tokenBridge/tokenBridge.ts#L231){target=\_blank} in [`tokenBridge.ts`](https://github.com/wormhole-foundation/wormhole-sdk-ts/blob/main/core/definitions/src/protocols/tokenBridge/tokenBridge.ts){target=\_blank}, abstracting away direct contract calls

```ts
// Assumes you have a tokenBridge instance from tokenBridge.ts set up

> const sender = "<SOURCE_CHAIN_SENDER_ADDRESS>";    // e.g., EVM address

Reviewer: ```suggestion
const sender = INSERT_SOURCE_CHAIN_SENDER_ADDRESS;    // e.g., EVM address
```

> const recipient = "<TARGET_CHAIN_RECIPIENT_ADDRESS>"; // e.g., Solana address in wormhole format

Reviewer: ```suggestion
const recipient = INSERT_TARGET_CHAIN_RECIPIENT_ADDRESS; // e.g., Solana address in wormhole format
```

> const tokenAddress = "<SOURCE_CHAIN_TOKEN_ADDRESS>";

Reviewer: ```suggestion
const tokenAddress = INSERT_SOURCE_CHAIN_TOKEN_ADDRESS;
```

const amount = BigInt("1000"); // Example amount

// 1. Initiate the transfer on the source chain:
for await (const tx of tokenBridge.transfer(sender, recipient, tokenAddress, amount)) {
  await sendTransaction(tx);
}

// 2. After the Guardians sign the VAA, obtain it from a Wormhole Guardian network source:
const transferVAA = ... // Obtain from Guardian network

// 3. On the destination chain, redeem the tokens:

> const receiver = "<YOUR_DESTINATION_CHAIN_ADDRESS>"; 

Reviewer: ```suggestion
const receiver = INSERT_YOUR_DESTINATION_CHAIN_ADDRESS; 
```

### Transferring Tokens with a Payload (Contract Controlled Transfers)

While a standard token transfer moves tokens between chains, a transfer with a payload allows you to embed arbitrary data in the VAA. This data can be used on the destination chain to execute additional logic—such as automatically depositing tokens into a DeFi protocol, initiating a swap on a DEX, or interacting with a custom smart contract.

Transferring tokens with payload flow:

1. **Source chain**
    - Call `ITokenBridge.transferTokensWithPayload()` instead of `transferTokens()`
    - Include a custom payload (arbitrary bytes) with the token transfer
2. **Guardian Network** - as with any transfer, the Guardians sign the VAA produced by the Token Bridge
3. **Destination chain**
    - On redemption, call `ITokenBridge.completeTransferWithPayload()` instead of `completeTransfer()`
    - Only the designated recipient contract can redeem these tokens. This ensures the attached payload is securely handled by the intended contract

Relevant methods and code references:

> - **Source Chain Initiation** - `ITokenBridge.transferTokensWithPayload()` is defined in [`ITokenBridge.sol`](https://github.com/wormhole-foundation/wormhole-solidity-sdk/blob/main/src/interfaces/ITokenBridge.sol#L101){target=\_blank}. Underlying logic for logging these payload-carrying transfers can be found in [`Bridge.sol` (`logTransferWithPayload`)](https://github.com/wormhole-foundation/wormhole/blob/main/ethereum/contracts/bridge/Bridge.sol#L336){target=\_blank}

Reviewer: ```suggestion
- **Source Chain Initiation** - `ITokenBridge.transferTokensWithPayload()` is defined in [`ITokenBridge.sol`](https://github.com/wormhole-foundation/wormhole-solidity-sdk/blob/main/src/interfaces/ITokenBridge.sol#L101){target=\_blank}. You can find the underlying logic for logging these payload-carrying transfers in [`Bridge.sol` (`logTransferWithPayload`)](https://github.com/wormhole-foundation/wormhole/blob/main/ethereum/contracts/bridge/Bridge.sol#L336){target=\_blank}
```

- **Destination chain redemption** - [`ITokenBridge.completeTransferWithPayload()`](https://github.com/wormhole-foundation/wormhole-solidity-sdk/blob/main/src/interfaces/ITokenBridge.sol#L114){target=\_blank} ensures that only the intended recipient address can redeem the tokens and process the payload.
- **SDK Integration** - The Wormhole SDK provides a single [`tokenBridge.transfer()`](https://github.com/wormhole-foundation/wormhole-sdk-ts/blob/main/core/definitions/src/protocols/tokenBridge/tokenBridge.ts#L215){target=\_blank} method that can optionally take a payload parameter. If provided, the SDK uses `transferTokensWithPayload` under the hood. Likewise, redemption calls `completeTransferWithPayload()` when it detects a payload, which is handled by [`tokenBridge.redeem()`](https://github.com/wormhole-foundation/wormhole-sdk-ts/blob/main/core/definitions/src/protocols/tokenBridge/tokenBridge.ts#L231){target=\_blank}

```ts
// Similar setup to a normal transfer, but we include a payload

> const sender = "<SOURCE_CHAIN_SENDER_ADDRESS>";

Reviewer: ```suggestion
const sender = INSERT_SOURCE_CHAIN_SENDER_ADDRESS;
```

> const recipient = "<TARGET_CHAIN_RECIPIENT_ADDRESS>";

Reviewer: ```suggestion
const recipient = INSERT_TARGET_CHAIN_RECIPIENT_ADDRESS;
```

> const tokenAddress = "<SOURCE_CHAIN_TOKEN_ADDRESS>";

Reviewer: ```suggestion
const tokenAddress = INSERT_SOURCE_CHAIN_TOKEN_ADDRESS;
```

const amount = BigInt("50000"); // Example amount
const customPayload = new Uint8Array([0x01, 0x02, 0x03]); // Arbitrary data

// 1. Initiate a transfer with payload on the source chain:
for await (const tx of tokenBridge.transfer(sender, recipient, tokenAddress, amount, customPayload)) {
  await sendTransaction(tx);
}

// 2. After obtaining the payload-carrying VAA from the Guardians:
const payloadVAA = ... // obtained from Wormhole Guardian network

// 3. On the destination chain, redeem:

> const receiver = "<DESTINATION_CHAIN_CONTRACT_ADDRESS>";

Reviewer: ```suggestion
const receiver = INSERT DESTINATION_CHAIN_CONTRACT_ADDRESS;
```

3. **Execution and token delivery** - on successful redemption, the tokens are minted (if foreign) or released (if native) to the recipient address on the destination chain. For payload transfers, the designated contract can execute the payload’s logic at this time

Relevant methods and code references:

- **Redemption functions** - `ITokenBridge.completeTransfer()` and `ITokenBridge.completeTransferWithPayload()` are both defined in [`ITokenBridge.sol`](https://github.com/wormhole-foundation/wormhole-solidity-sdk/blob/main/src/interfaces/ITokenBridge.sol){target=\_blank}. On-chain logic for processing these calls resides in [`Bridge.sol`](https://github.com/wormhole-foundation/wormhole/blob/main/ethereum/contracts/bridge/Bridge.sol){target=\_blank}
- **SDK integration** - The Wormhole SDK’s `tokenBridge.redeem()` method (in [`tokenBridge.ts`](https://github.com/wormhole-foundation/wormhole-sdk-ts/blob/main/core/definitions/src/protocols/tokenBridge/tokenBridge.ts){target=\_blank}) accepts the VAA and automatically handles calling the correct function on-chain

```ts
const transferVAA = ...; // Obtained from Guardian network

> const receiver = "<DESTINATION_CHAIN_ADDRESS>";

Reviewer: ```suggestion
const receiver = INSERT_DESTINATION_CHAIN_ADDRESS;
```

This page will walk you through the essential methods and events of the Token Bridge contracts, providing you with the knowledge needed to integrate them into your cross-chain applications. For more details on how the Token Bridge works, refer to the [Token Bridge](/docs/learn/messaging/token-nft-bridge/){target=\_blank} or [Native Token Transfers](/docs/learn/messaging/native-token-transfers/overview/#token-bridge){target=\_blank} pages in the Learn section and the [Token Bridge Whitepaper](https://github.com/wormhole-foundation/wormhole/blob/main/whitepapers/0003_token_bridge.md){target=\_blank}.

## Prerequisites

To interact with the Wormhole Token Bridge, you'll need the following:

> - **SDK Installation** - Ensure the Wormhole SDK is installed in your project. You can install it via npm or yarn:

Reviewer 2: ```suggestion
- **SDK installation** - ensure the Wormhole SDK is installed in your project. You can install it via npm or yarn:
```

Reviewer 2: Why do they need the SDK? This guide should cover contract to contract interactions, like the Core Contract and CCTP pages

> - **SDK Installation** - Ensure the Wormhole SDK is installed in your project. You can install it via npm or yarn:

Reviewer 2: These list items kinda look weird cause of the different formatting. Some lines you summarize with the text in bold, others you don't. I'd just prefer them all to be the same format

```sh
npm install @wormhole-foundation/sdk
```
- [The address of the Token Bridge Core Contract](/docs/build/reference/contract-addresses#core-contracts) on the chains you're working with
- [The Wormhole chain ID](/docs/build/reference/chain-ids/) of the chains you're you're targeting for token transfers

> - **Token attestation** - the token you wish to transfer must be compatible with the destination chain as a wrapped asset. For tokens that do not already exist on the target chain, you must attest their details, which includes metadata like `name`, `symbol`, `decimals`, and `payload_id`, which must be set to `2` for an attestation. This ensures that the wrapped token on the destination chain preserves the original token’s properties. The attestation process records the token's metadata on the target chain to ensure its consistency across chains. See the [Attestation section](/docs/learn/infrastructure/vaas/#attestation){target=\_blank} for more details and all the required metadata fields

Reviewer 2: ```suggestion
- **Token attestation** - the token you wish to transfer must be compatible with the destination chain as a wrapped asset. For tokens that do not exist on the target chain, you must attest their details, including metadata like `name`, `symbol`, `decimals`, and `payload_id`, which must be set to `2` for an attestation. This ensures that the wrapped token on the destination chain preserves the original token’s properties. The attestation process records the token's metadata on the target chain to ensure consistency across chains. See the [Attestation section](/docs/learn/infrastructure/vaas/#attestation){target=\_blank} for more details and all the required metadata fields
```

## How to Interact with the Token Bridge Contracts

> The Wormhole Token Bridge SDK offers a set of TypeScript types and functions that make it easy to interact with the bridge. This includes handling token transfers, verifying wrapped assets, and submitting attestations. The key functionality revolves around:
> 
> - **Token transfers** - locking tokens on the source chain and minting wrapped tokens on the destination chain
> - **Attestations** - submitting metadata about a token to other chains for wrapping
> - **Verifying wrapped assets** - checking if a token is wrapped and retrieving its original chain and asset details

Reviewer 2: ```suggestion
The Wormhole Token Bridge SDK offers a set of TypeScript types and functions that make it easy to interact with the bridge. The key functionality revolves around:

- **Token transfers** - locking tokens on the source chain and minting wrapped tokens on the destination chain
- **Attestations** - submitting metadata about a token to other chains for wrapping
- **Verifying wrapped assets** - checking if a token is wrapped and retrieving its original chain and asset details
```

> ### Check if a Token is Wrapped

Reviewer 2: ```suggestion
### Check If a Token Is Wrapped
```

## Prerequisites

To interact with the Wormhole Token Bridge, you must ensure you have the addresses and chain IDs for the Wormhole Core and Token Bridge contracts on the networks you want to work with.

- [The address of the Token Bridge Core Contract](/docs/build/reference/contract-addresses/#token-bridge) on the chains you're working with
- [The Wormhole chain ID](/docs/build/reference/chain-ids/) of the chains you're you're targeting for token transfers

## Setup and Installation

Before you can start interacting with Wormhole's Token Bridge, you need to set up your development environment by installing the Wormhole SDK and configuring your project to use it. This section guides you through the necessary steps to get started.

1. **Install the Wormhole SDK** 
  

>     First, ensure that you have `Node.js` and `npm` installed on your machine. Then, install the Wormhole SDK packages:

Reviewer: ```suggestion
    First, ensure you have `Node.js` and `npm` installed on your machine. Then, install the Wormhole SDK packages:
```

    ```sh
    npm install @wormhole-foundation/wormhole-sdk @wormhole-foundation/sdk-base
    ```

    - `@wormhole-foundation/wormhole-sdk` - core Wormhole SDK for interacting with the Wormhole network
    - `@wormhole-foundation/sdk-base` - Base SDK provides foundational types and interfaces

2.  **Import the necessary modules**

    In your TypeScript or JavaScript file (e.g., index.ts), import the required classes and functions from the SDK:

    ```ts
    import { TokenBridge, createTokenBridge } from '@wormhole-foundation/sdk-base';
    import { Relayer, createRelayer } from '@wormhole-foundation/sdk-base'; // If using Relayer
    // Import other necessary modules as needed
    ```

3. **Initialize the TokenBridge Interface**

>     To use methods like `createAttestation()`, `transfer()`, `redeem()`, and other, you need to instantiate the respective interfaces provided by the SDK. Here's how to set it up in a blockchain-agnostic manner:

Reviewer: ```suggestion
    To use methods like `createAttestation(),` `transfer(),` `redeem(),` and others, you need to instantiate the respective interfaces provided by the SDK. Here's how to set it up in a blockchain-agnostic manner:
```

!!!note
    The code snippets below are simplified and focus on the main calls. For full implementations, refer to the provided contract source files (like [`bridge/Bridge.sol` ](https://github.com/wormhole-foundation/wormhole/blob/main/ethereum/contracts/bridge/Bridge.sol){target=\_blank} and [`ITokenBridge.sol`](https://github.com/wormhole-foundation/wormhole-solidity-sdk/blob/main/src/interfaces/ITokenBridge.sol){target=\_blank}) and your integrated SDK code (e.g., `tokenBridge.ts`).
    
    The examples assume you have set up a project with TypeScript, the Wormhole SDK, RPC endpoints, and private keys configured.

### Attesting a Token

If a token has never been transferred to a target chain, it must be attested so the Token Bridge can recognize its metadata (decimals, name, symbol) and create a wrapped version if necessary.

Attesting a new token involves providing its metadata and setting the `payload_id` to `2`. This ensures the wrapped token on the destination chain preserves the original token's properties and consistency across chains. For more details, see the [Attestation section](/docs/learn/infrastructure/vaas/#attestation){target=\_blank}.

> The attestation process doesn’t require you to manually input token details like name, symbol, or decimals. The Token Bridge contract retrieves these values from the token contract itself when you call the `attestToken()` method.

Reviewer: ```suggestion
The attestation process doesn't require you to manually input token details like name, symbol, or decimals. The Token Bridge contract retrieves these values from the token contract itself when you call the `attestToken()` method.
```
